### DISCUSSION TITLE
[QUESTION] Change cross section using IdeaStatiCa.Plugin;

### DISCUSSION THREAD

**Discussion by ZTThomasOberthanner** [idea-statica/ideastatica-public#459] [Help Q&A] [answered]
I am working on a small command line tool that loads a template.ideaCon file containing a single connection using c# and IdeaStatiCa.Plugin with three goals:

1. update loading
2. update parameters
3. update cross sections

My whole project is based on [examples/api/csharp/connection/Connection-API-Console-Tester/ConHiddenCheckConsole](https://github.com/idea-statica/ideastatica-public/tree/main/examples/api/csharp/connection/Connection-API-Console-Tester/ConHiddenCheckConsole)

My question:
How can I change the cross section of that connection? (updating of loading and parameters already works :-) )

As the api documentation is yet not complete I had a look in the examples and found out, that I can use CrossSectionFactory inside IOM to create a cross section. As I understand it, it is not possible to use library cross sections like HEB200. But anyway, I am already stuck on how to tell the client to assign the cross section to the connection. Any ideas?

```
// create the instance of factory - it looks for IDEA StatiCa in the directory 'ideaStaticaInstallDir'
var calcFactory = new ConnHiddenClientFactory(IDEASTATICADIR);

//create the instance of the IDEA StatiCa Client
ConnectionHiddenCheckClient client = calcFactory.Create();

client.OpenProject(ideaProjectFileName);
// get detail about idea connection project
IdeaRS.OpenModel.Connection.ConProjectInfo projInfo = client.GetProjectInfo();

// get first connection
IdeaRS.OpenModel.Connection.ConnectionInfo con = projInfo.Connections.First();

// update cross sections
CrossSectionParameter crossSectionParameter = new CrossSectionParameter() { Id = 1, Name = "myISection" };
CrossSectionFactory.FillRolledI(crossSectionParameter, 0.25, .045, 0.02, 0.03, 0.012, 0, 0.015);

// update parameter
client.ApplyParameters(con.Identifier, paramArray.ToString());

// update loading
client.UpdateLoadingFromJson(con.Identifier, template.Loadcases.ToString());
```

As a workaround I could also use a generic I in the template and set height, width, thickness of flanges and so on using parameters.

**Comment by ZTThomasOberthanner** [comment]
Second question: how to change the material for a cross section?

**Comment by ZTThomasOberthanner** [comment]
Third question: how to change a bolt dimension and quality (welds are working using parameters)

**Comment by nathanlukeIDEA** [comment, marked as answer]
Hi @ZTThomasOberthanner,

From what I gather you are:
1.0 Opening an existing file
2.0 Making some changes: Loading, Section Sizes, Material, Bolt Size.
3.0 Recalculating the connection.

For changing Items. I believe all of what you are asking for is also available through parameters and this would be the best way to perform it. 

1.0 Let's say you have these library items available in your project:
![image](https://github.com/idea-statica/ideastatica-public/assets/93578548/0610cc6e-dad0-4c05-99b3-7466fcab8dc0)

2.0 You can use these API calls to get the available item names and ids in the existing project This can help you drive the parameter updates when wanting to select the section, material etc: 
`GetCrossSectionsInProject()`
`GetBoltAssembliesInProject()`
`GetMaterialsInProject()`

> How can I change the cross section of that connection? (updating of loading and parameters already works :-) )

Given you have a list of cross-sections in the project you can change a member to any one of these using parameters. 
![image](https://github.com/idea-statica/ideastatica-public/assets/93578548/51ee68ff-be8b-4c6e-91cb-9d88a9171fc0)

> Second question: how to change the material for a cross section?

![image](https://github.com/idea-statica/ideastatica-public/assets/93578548/c0450eec-4674-4767-bc0e-471eb3f421ee)

> Third question: how to change a bolt dimension and quality (welds are working using parameters)
![image](https://github.com/idea-statica/ideastatica-public/assets/93578548/0bc35d28-af92-46a9-a28d-d0579a512a49)

You can then simply use the update parameters call to make all these changes at once.

I hope this helps, let me know if you require anything else.

Cheers
Nathan

**Reply by ZTThomasOberthanner** [reply]
Thank you for the detailed reply. As I understand this implies that it is required to add
- all possible cross sections
- all required bolt size and quality combinations
- all required materials
to the template. This is possible for materials and roughly 20 bolt combinations but it is just not possible to add >50 I sections, >200 hollow sections and so on manually to the template. Is there a quick way to automatically add ALL available cross sections from the library to the project? I guess that is not possible so I will take the other approach (generic I section and set dimensions and flange thickness using parameters). This is a little bit cumbersome to set up in the template but way easier to code and test.

**Reply by nathanlukeIDEA** [reply]
Hi @ZTThomasOberthanner, The other way to do it is to have a parameter set to a named cross-section.  

In this instance the when you set the cross-section it will search the database and change the cross-section based on the provided library name. Note it will change any member that references the cross-section. You also need to be careful that the cross-section shape name is correct with the IDEA library. 

![image](https://github.com/idea-statica/ideastatica-public/assets/93578548/74c649eb-ebf8-419a-9ea1-9d50007e8d24)



You would need to have separate cross-sections for different types like I, or Hollow sections etc.

**Reply by ZTThomasOberthanner** [reply]
Hi @nathanlukeIDEA 
That is exactly what I was looking for. The advantage of that solution is that the cross section is correctly labeled in the report. Will try that approach. Thank you for your help!

**Reply by nathanlukeIDEA** [reply]
Great, let me know how it goes! Cheers

**Reply by ZTThomasOberthanner** [reply]
Updating cross section/material and creating load cases already works. Importing the results from idea as well. will have a look at the bolts today, the rest (geometry) should be a walk in the park...

@nathanlukeIDEA Is it safe to run idea in parallel? Because my own steel connection tool calculates the steel checks and creates drawings and reports in parallel using multiple cores. And as I now have to possibility to use idea as a backend for some connection types where cbfem is better than the ec3 equations this would be nice.

**Reply by ZTThomasOberthanner** [reply]
<img width="497" alt="image" src="https://github.com/idea-statica/ideastatica-public/assets/92015468/14d0be90-7584-443e-8275-a1a67430b264">

is there a way to set the material of the headplate? I can only spot two rows material for the welds but not for the head plate itself..

**Reply by nathanlukeIDEA** [reply]
It should be avaliable, we are looking into this. I have created an issue: https://github.com/idea-statica/ideastatica-public/issues/463

**Comment by nathanlukeIDEA** [comment]
> Updating cross section/material and creating load cases already works. Importing the results from idea as well. will have a look at the bolts today, the rest (geometry) should be a walk in the park...
> 
> @nathanlukeIDEA Is it safe to run idea in parallel? Because my own steel connection tool calculates the steel checks and creates drawings and reports in parallel using multiple cores. And as I now have to possibility to use idea as a backend for some connection types where cbfem is better than the ec3 equations this would be nice.

@ZTThomasOberthanner Please see this thread and check if it answers your query. https://github.com/idea-statica/ideastatica-public/discussions/151#discussioncomment-2740689

**Reply by ZTThomasOberthanner** [reply]
@nathanlukeIDEA thanks for the updates.
#151 tells me that running parallel should be safe. And a quick look at the CPU utilization tells me that the performance should be better (report generation seems to be single threaded). 

<img width="698" alt="image" src="https://github.com/idea-statica/ideastatica-public/assets/92015468/a087b5f8-7e29-4a87-96e6-320f85d12c6a">
